### PR TITLE
[draft] Aten cat to tosa

### PR DESCRIPTION
**Pull request by yifei410**

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -4655,6 +4679,7 @@ class ConvertTorchToTosa : public ConvertTorchToTosaBase<ConvertTorchToTosa> {
     INSERT_ATENOP_PATTERN(AtenToDtypeOp);
     INSERT_ATENOP_PATTERN(AtenConstantPadNdOp);
     INSERT_ATENOP_PATTERN(AtenRemainderScalarOp);
+    INSERT_ATENOP_PATTERN(AtenCatOp);
```
Comment (ljfitz) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4689]: Wondering why tests like [TensorsConcatModule](https://github.com/llvm/torch-mlir/blob/a13d301356fe8fe414a707ed3a3e011cd888e3bb/python/torch_mlir_e2e_test/test_suite/basic.py#L576) fail. If it must fail (I guess because the tensors have dynamic shapes) then you probably need to replicate the existing `cat` tests to make `Static` versions as you see elsewhere in `basic.py` for other operators.

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 2]
```diff
@@ -647,6 +647,29 @@ def TensorsConcatPromoteDTypeModule_basic(module, tu: TestUtils):
 # ==============================================================================
 
 
+class CatModule(torch.nn.Module):
```
Comment (ljfitz) [inline python/torch_mlir_e2e_test/test_suite/basic.py]: Consistency in naming is important, especially with closely related tests. Effectively I believe you are testing `torch.cat`, which gets decomposed to `torch.ops.aten.cat`, so this test should follow the naming scheme of the `torch.cat` operation. If you think the naming there is inconsistent, then you should rename them all (though in my first PR I'd probably add a code comment in my own PR with a question to reviewers to know if the renaming would be appreciated).

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 3]
```diff
@@ -647,6 +647,29 @@ def TensorsConcatPromoteDTypeModule_basic(module, tu: TestUtils):
...
+
+    @export
+    @annotate_args([
+        None,
+        ([2, 2, 4], torch.float32, True),
+        ([2, 1, 4], torch.float32, True),
+        ([2, 3, 4], torch.float32, True),
+    ])
+    def forward(self, x, y, z):
+        return torch.ops.aten.cat([x, y, z], dim=1)
```
Comment (ljfitz) [inline python/torch_mlir_e2e_test/test_suite/basic.py]: Unless it really has a different result, please use `torch.cat` instead of `torch.ops.aten.cat`.

**Review by ljfitz — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (yifei410) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4682]: for tosa I could only get positive dim to work(https://mlir.llvm.org/docs/Dialects/TOSA/#tosaconcat-mlirtosaconcatop), so just have one static positive dim test. I have renamed the test as you suggested

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -4430,6 +4430,30 @@ LogicalResult ConvertAtenOp<AtenConstantPadNdOp>::matchAndRewrite(
   return success();
 }
 
+template <>
+LogicalResult ConvertAtenOp<AtenCatOp>::matchAndRewrite(
+    AtenCatOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  int64_t dim;
+  if (!matchPattern(op.getDim(), m_TorchConstantInt(&dim)))
```
Comment (lisaliu1) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: can you check if the dim is >= 0 since tosa doesn't support signed number for the dim argument?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (yifei410) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4689]: just add support for negative dim in the cpp side and using torch.cat now

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 4 above)
Comment (yifei410) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: just add support for negative dim

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 5]
```diff
@@ -647,6 +647,54 @@ def TensorsConcatPromoteDTypeModule_basic(module, tu: TestUtils):
 # ==============================================================================
 
 
+class ConcatStaticModule(torch.nn.Module):
```
Comment (lisaliu1) [inline python/torch_mlir_e2e_test/test_suite/basic.py]: can you change this class name to "TensorConcatStaticModule"? So it's aligned with other test names. Thanks.

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 6]
```diff
@@ -647,6 +647,54 @@ def TensorsConcatPromoteDTypeModule_basic(module, tu: TestUtils):
...
+        ([2, 2, 4], torch.float32, True),
+        ([2, 1, 4], torch.float32, True),
+        ([2, 3, 4], torch.float32, True),
+    ])
+    def forward(self, x, y, z):
+        return torch.cat([x, y, z], dim=1)
+
+
+@register_test_case(module_factory=lambda: ConcatStaticModule())
+def ConcatStaticModule_basic(module, tu: TestUtils):
```
Comment (lisaliu1) [inline python/torch_mlir_e2e_test/test_suite/basic.py]: Can you change this test case name to "TensorConcatStaticModule_basic"?

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 7]
```diff
@@ -647,6 +647,54 @@ def TensorsConcatPromoteDTypeModule_basic(module, tu: TestUtils):
...
+    module.forward(tu.rand(2, 2, 4), tu.rand(2, 1, 4), tu.rand(2, 3, 4))
+
+
+# ==============================================================================
+
+
+class ConcatNegativeDimStaticModule(torch.nn.Module):
```
Comment (lisaliu1) [inline python/torch_mlir_e2e_test/test_suite/basic.py]: Can you change this class name to "TensorConcatNegativeDimStaticModule" and the test case to "TensorConcatNegativeDimStaticModule_basic"? Please also change the ones in TOSA_PASS_SET as well. Thanks.

**Review by ljfitz — APPROVED** [review]
I agree with the comments from @lisaliu1 but otherwise LGTM. Thanks for taking on this upstreaming effort.
